pkg/apis/cloudcredential/v1/credentialsrequest_types.go

```diff
@@ -44,17 +44,55 @@ const (
 	CloudCredOperatorConfigMap = "cloud-credential-operator-config"
 )
 
-// NOTE: Run "make" to regenerate code after modifying this file
+// NOTE: Run "make update" to regenerate code after modifying this file
 
 // CredentialsRequestSpec defines the desired state of CredentialsRequest
 type CredentialsRequestSpec struct {
 	// SecretRef points to the secret where the credentials should be stored once generated.
-	SecretRef corev1.ObjectReference `json:"secretRef"`
+	// +optional
+	SecretRef corev1.ObjectReference `json:"secretRef,omitempty"`
+
+	// StorageDestination specifies the type and location of a resource to which the credentials
+	// information can be stored or linked
+	StorageDestination StorageDestination `json:"storage,omitempty"`
```

Comment:
Apologies I may have misunderstood your other proposal (3), but I expected StorageDestination here to be more like StorageType, which is a const string just indicating which field is set on the CredentialsRequest. I don't think we should introduce a new way to set the secret target when we have to keep the old.

i.e.
credRequest.Spec.Type or StorageType
credRequest.Spec.SecretRef
credRequest.Spec.ServiceAccountRef

Reply:
We could flatten the `storage` field down into the `spec`.  The reason I didn't is that now we have to deal with the existing `SecretRef` at that level and if we continue to use that we can't deprecate it in favor of the better typed `secret` field in the `storage` union.

Reply:
I'm +1 for collapsing, I really don't want two secretRef fields with one deprecated on our v1 API, immediately making all our cred requests today in various repos technically using deprecated fields.

I'm +1 for ServiceAccountRef being ObjectReference, keeping it consistent with SecretRef, even though this is not recommended use now. +0 for ServiceAccountRef being typed instead of ObjectRef, it's weird to have Secret and ServiceAccount out of sync but I can live with it.

I'm +1 for making SecretRef typed as well *if* that's ok, from an API caller perspective I think it's valid as we're still just using name+namespace but I have no idea if this is an acceptable thing to do. I'm going to ask in forum-api-review.

Reply:
I recommend against collapsing the discriminated union in your top level spec.  It's easier to reason about what an API does if the "pick one" isn't co-located with "applies in to anything" (in this case your `providerSpec` for instance).

Reply:
The one thing I hoped to avoid in this is two places to set the secret ref, deprecating all our existing usage, and carrying that confusion forever.